File: skills/asset-validation/scripts/cleanup.py
from pathlib import Path

from . import db, observe, rounds


def cleanup_round(con, round_id, *, sandbox=None) -> dict:
    row = rounds.get_cleanup_target(con, round_id)
    if not row:
        raise LookupError(f"round not found: {round_id}")

    sandbox = sandbox or row["sandbox_path"]
    if not sandbox:
        raise ValueError("cleanup requires --sandbox or --round")

    session = observe.session_name(row["round_tag"])
    session_killed = observe.kill_session(session)
    nested_sessions = _cleanup_nested_sessions(sandbox)
    plugin_cleanup = None
    if row["asset_type"] in ("plugin", "skill", "agent"):
        plugin_cleanup = observe.cleanup_plugin_install(sandbox)

    result = observe.cleanup(sandbox)
    result.update({"session": session, "session_killed": session_killed})
    if nested_sessions:
        result["nested_sessions"] = nested_sessions
    if plugin_cleanup is not None:
        result["plugin_cleanup"] = plugin_cleanup
    return result


def _cleanup_nested_sessions(sandbox) -> list:
    state_db = Path(sandbox) / ".aut-acceptance" / "state.sqlite3"
    cleaned = []
    for target in db.round_cleanup_targets_from(state_db):
        round_tag = target.get("round_tag")
        if not round_tag:
            continue
        session = observe.session_name(round_tag)
        cleaned.append({
            "session": session,
            "killed": observe.kill_session(session),
        })
    return cleaned

File: skills/asset-validation/scripts/cleanup.ts
import { join } from "node:path";

import * as db from "./db.ts";
import { LookupError, ValidationError } from "./errors.ts";
import * as observe from "./observe.ts";
import * as rounds from "./rounds.ts";

export function cleanupRound(
  connection: db.Connection,
  roundId: string,
  options: { sandbox?: string | null } = {},
): Record<string, unknown> {
  const row = rounds.getCleanupTarget(connection, roundId);
  if (!row) throw new LookupError(`round not found: ${roundId}`);
  const sandbox = options.sandbox || row.sandbox_path;
  if (!sandbox) throw new ValidationError("cleanup requires --sandbox or --round");
  const session = observe.sessionName(row.round_tag);
  const sessionKilled = observe.killSession(session);
  const nestedSessions = cleanupNestedSessions(sandbox);
  let pluginCleanup: Record<string, unknown> | null = null;
  if (["plugin", "skill", "agent"].includes(row.asset_type)) {
    pluginCleanup = observe.cleanupPluginInstall(sandbox);
  }
  const result = observe.cleanup(sandbox);
  Object.assign(result, { session, session_killed: sessionKilled });
  if (nestedSessions.length) result.nested_sessions = nestedSessions;
  if (pluginCleanup !== null) result.plugin_cleanup = pluginCleanup;
  return result;
}

function cleanupNestedSessions(sandbox: string): Array<Record<string, unknown>> {
  const stateDatabase = join(sandbox, ".aut-acceptance", "state.sqlite3");
  const cleaned: Array<Record<string, unknown>> = [];
  for (const target of db.roundCleanupTargetsFrom(stateDatabase)) {
    if (!target.round_tag) continue;
    const session = observe.sessionName(target.round_tag);
    cleaned.push({ session, killed: observe.killSession(session) });
  }
  return cleaned;
}
